src/crypto-providers/signingValidation.ts

```diff
@@ -0,0 +1,232 @@
+  // expected to be complete) on top of validateOrdinaryWitnesses
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles, mintSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length === 0) {
+    throw Error(Errors.MissingPaymentSigningFileError)
+  }
+
+  const {
+    numStakeWitnesses, numPoolColdWitnesses,
+  } = _countWithdrawalAndCertificateWitnesses(params.unsignedTxParsed)
+  const numMintWitnesses = params.unsignedTxParsed.mint?.length || 0
+
+  if (numStakeWitnesses > 0 && (stakeSigningFiles.length === 0)) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (numPoolColdWitnesses > 0 && (poolColdSigningFiles.length === 0)) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+  if (numMintWitnesses > 0 && (mintSigningFiles.length === 0)) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+}
+
+const validatePoolRegistrationTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateNonMultisigTx(unsignedTxParsed)
+
+  if (unsignedTxParsed.inputs.length === 0) {
+    throw Error(Errors.MissingInputError)
+  }
+  if (unsignedTxParsed.certificates.length !== 1) {
+    throw Error(Errors.MultipleCertificatesWithPoolRegError)
+  }
+  if (unsignedTxParsed.withdrawals.length > 0) {
+    throw Error(Errors.WithdrawalIncludedWithPoolRegError)
+  }
+}
+
+const validatePoolOwnerWitnesses = (params: SigningParameters) => {
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles, mintSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length > 0) {
+    throw Error(Errors.TooManyPaymentFilesWithPoolRegError)
+  }
+  if (stakeSigningFiles.length === 0) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (stakeSigningFiles.length > 1) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (poolColdSigningFiles.length > 0) {
+    throw Error(Errors.TooManyPoolColdSigningFilesError)
+  }
+  if (mintSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validatePoolOperatorWitnesses = (params: SigningParameters) => {
+  const {
+    stakeSigningFiles, poolColdSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (poolColdSigningFiles.length === 0) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+  if (stakeSigningFiles.length > 0) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validateMultisigTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  // We require multisig transactions to use certificates and withdrawals
+  // only with ScriptHash stake credentials
```

Comment:
I'd suggest to extract the check into a function `_validateScriptStakeCredentialsEverywhere` analogous to `_validateKeyStakeCredentialsEverywhere` (see the suggestion above) and just call it here, then we won't need the comment.

Reply:
Resolved by the `_validateStakeCredentialsAreAllOfType` function (above).